ignite/metrics/nlp/bleu.py

```diff
@@ -122,14 +126,105 @@ def __init__(
         smooth: str = "no_smooth",
         output_transform: Callable = lambda x: x,
         device: Union[str, torch.device] = torch.device("cpu"),
+        average: str = "macro",
     ):
         if ngram <= 0:
             raise ValueError(f"ngram order must be greater than zero (got: {ngram})")
         self.ngrams_order = ngram
         self.weights = [1 / self.ngrams_order] * self.ngrams_order
         self.smoother = _Smoother(method=smooth)
+
+        if average not in ["macro", "micro"]:
+            raise ValueError(f'Average must be either "macro" or "micro" (got: {average})')
+        self.average = average
+        self.p_numerators: Counter = Counter()
+        self.p_denominators: Counter = Counter()
+        self.hyp_length_sum = 0
+        self.ref_length_sum = 0
+
         super(Bleu, self).__init__(output_transform=output_transform, device=device)
 
+    def _n_gram_counter(
```

Comment:
This method should be refactored to be static. It means pass `p_denominators`, `p_numerators` and lengths as arguments. IMO it would help to design the rest of the code following the idea below.
```python
@static
def _n_gram_counter(...):
   ...

@static
def _brevity_penalty_smoothing(...):
    ...

# implements the corpus bleu algorithm using `_n_gram_counter` and `_brevity_penalty_smoothing`
# i.e. the candidate input is a list of sentences
def _corpus_bleu(...) -> float:
    ...

# implements the sentence bleu algorithm using `_corpus_bleu` as nltk
# i.e. the candidate input is a sentence
def _sentence_bleu(...) -> float:
    ...
```

The `_sentence_bleu` and `_corpus_bleu` should be compared to nltk.

What do you think ?

Reply:
Edit : I missed that we used some parameters (smoother, etc.). Having static methods is quite strong, involving a plenty of arguments in the signature. Anyway, the counters and lengths must be in the signature (or returned by the methods). Sorry for this.